### PR TITLE
Add symfony shutdown strategy

### LINKED ISSUE #77: register_shutdown_function() causes major memory leak during testing
### Expected behavior
Symfony tests should run under phpunit without major memory leaks.

Alternatively, provide a bootstrapping process which can be modified for testing. Currently, the hard-coding of the `Bugsnag\Client` class makes that very difficult.

### Observed behavior
Each request/sub-request during testing adds 2MB-4MB of permanent memory consumption.

### Steps to reproduce
```
class TestCase extends \Symfony\Bundle\FrameworkBundle\Test\WebTestCase 
{
    public function testMemory()
    {
        echo memory_get_usage() . PHP_EOL;
        $client->request('GET', '/');
        echo memory_get_usage() . PHP_EOL;
        $client->request('GET', '/');
        $client->request('GET', '/');
        $client->request('GET', '/');
        $client->request('GET', '/');
        $client->request('GET', '/');
        echo memory_get_usage() . PHP_EOL;
    }
}

```
### Version
1.4.0/1.5.0

### Additional information
I have confirmed that removing the `register_shutdown_function([$this, 'flush']);` from the `Client` constructor resolves the issue. Unfortunately, that encloses the state of `$this` until the process shuts down, which on large test suites is usually when it crashes from memory problems.

### PR DESCRIPTION
**Pull request by rjharrison**
## Goal

Adds a Symfony-specific shutdown strategy. This replaces the default PhpShutdownStrategy (`register_shutdown_function`) with a Symfony event-subscriber that reacts to the `Kernel::TERMINATE` event. 

This is motivated by two benefits: 

1. No memory leaks when running tests: memory used by the `Bugsnag\Client` can be garbage collected (this is not possible when `register_shutdown_function` is used).

2. Better performance: Symfony uses `fastcgi_finish_request()`, so if PHP-FPM is being used, the `flush()` call takes place after the HTTP connection has closed.

## Design

Avoiding the [memory leak](https://github.com/bugsnag/bugsnag-symfony/issues/77) when testing was the primary motivation. Potentially we could have done this by implementing conditional registration of the shutdown strategy (i.e. only register it in batch mode – current behaviour is to always register it within the Client constructor). This would have allowed users to test without the memory leak by setting `batch_sending = false` for the test environment.

Whilst conditional registration of the shutdown strategy is still desirable (why register one if we are not batch sending?), a better solution is to take advantage of Symfony's `KernelEvents::TERMINATE` event. When using `WebTest` to perform functional tests, Symfony's kernel is rebooted between tests (i.e. `KernelEvents::TERMINATE` is fired after the test). This destroys the DI container and allows memory to be garbage collected by PHP. 

### Added

- BugsnagShutdown class (implements ShutdownStrategyInterface + EventSubscriberInterface)

### Changed

- ClientFactory now instantiates Bugsnag\Client with the new shutdown strategy
- services.yml has a new definition for bugsnag.shutdown + injection into ClientFactory

## Tests

- Added unit tests for new changes

### Alternative Approaches

As above, I considered @ryancastle's suggestion of removing the call to `register_shutdown_function` from the constructor (and calling it only when `batch_sending = true`).

### Outstanding Questions

I think we should consider conditionally registering (or deregistering) the shutdown strategy depending on whether we are batching or sending synchronously. 

### Linked issues

Fixes #77 

## Review

For the submitter, initial self-review:

- [x] Commented on code changes inline explain the reasoning behind the approach
- [x] Reviewed the test cases added for completeness and possible points for discussion
- [ ] A changelog entry was added for the goal of this pull request
- [x] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [ ] Initial review of the intended approach, not yet feature complete
  - [x] Structural review of the classes, functions, and properties modified
  - [ ] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

**Comment by rjharrison** [comment]
Builds are failing as this code requires the unmerged new stuff in https://github.com/bugsnag/bugsnag-php/pull/547

File: EventListener/BugsnagShutdown.php [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+    /**
+     * Indicate which events we wish to subscribe to.
+     *
+     * @return array
+     */
+    public static function getSubscribedEvents()
+    {
+        $listeners = [
+            KernelEvents::TERMINATE => ['onTerminate', 10],
```
[the remark this answers is not in the document]
Comment (rjharrison) [reply EventListener/BugsnagShutdown.php]: I think the event argument passed to the handler was changed (4.2 -> 4.3), but the constant remained the same. I'll make sure to double check tests pass on older Symfony versions.
